The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11779518752/job/32808162271) [(plain)](https://github.com/rust-lang/rust/commit/e87d73b8dbbd2f81e60ec41f549dd7673a08df1a/checks/32808162271/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:c32c805632780b5c1de330e3f44561b336c2efe163bc0990acb392390157a8e1d9f855d75914a239aa40c49d77f4a837247d05d2f8d46f554b98e1f46712a3e3:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
warning: MIR pass `DataflowConstProp` is unknown and will be ignored

warning: MIR pass `DataflowConstProp` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `DataflowConstProp` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `DataflowConstProp` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `DataflowConstProp` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: 5 warnings emitted





The actual stderr differed from the expected stderr.
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args codegen/const-bool-bitcast.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/codegen/const-bool-bitcast.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/codegen/const-bool-bitcast" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Zmir-opt-level=0" "-Zmir-enable-passes=+DataflowConstProp"
--- stderr -------------------------------
warning: MIR pass `DataflowConstProp` is unknown and will be ignored

warning: MIR pass `DataflowConstProp` is unknown and will be ignored
warning: MIR pass `DataflowConstProp` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `DataflowConstProp` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `DataflowConstProp` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `DataflowConstProp` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: 5 warnings emitted
------------------------------------------



---- [ui] tests/ui/debuginfo/sroa-fragment-debuginfo.rs stdout ----
Saved the actual stderr to "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/debuginfo/sroa-fragment-debuginfo/sroa-fragment-debuginfo.stderr"
normalized stderr:
warning: MIR pass `ScalarReplacementOfAggregates` is unknown and will be ignored

warning: MIR pass `ScalarReplacementOfAggregates` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `ScalarReplacementOfAggregates` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `ScalarReplacementOfAggregates` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `ScalarReplacementOfAggregates` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `ScalarReplacementOfAggregates` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `ScalarReplacementOfAggregates` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `ScalarReplacementOfAggregates` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `ScalarReplacementOfAggregates` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `ScalarReplacementOfAggregates` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: 10 warnings emitted





The actual stderr differed from the expected stderr.
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args debuginfo/sroa-fragment-debuginfo.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/debuginfo/sroa-fragment-debuginfo.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/debuginfo/sroa-fragment-debuginfo" "-A" "unused" "-A" "internal_features" "-Crpath" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-g" "-Zmir-opt-level=0" "-Zmir-enable-passes=+ScalarReplacementOfAggregates" "-Cno-prepopulate-passes"
--- stderr -------------------------------
--- stderr -------------------------------
warning: MIR pass `ScalarReplacementOfAggregates` is unknown and will be ignored

warning: MIR pass `ScalarReplacementOfAggregates` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `ScalarReplacementOfAggregates` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `ScalarReplacementOfAggregates` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `ScalarReplacementOfAggregates` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `ScalarReplacementOfAggregates` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `ScalarReplacementOfAggregates` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `ScalarReplacementOfAggregates` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `ScalarReplacementOfAggregates` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `ScalarReplacementOfAggregates` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: 10 warnings emitted
------------------------------------------



---- [ui] tests/ui/lint/invalid_value-polymorphic.rs stdout ----
Saved the actual stderr to "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/invalid_value-polymorphic/invalid_value-polymorphic.stderr"
normalized stderr:
warning: MIR pass `InstSimplify` is unknown and will be ignored

warning: MIR pass `InstSimplify` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `InstSimplify` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `InstSimplify` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `InstSimplify` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `InstSimplify` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: 6 warnings emitted





The actual stderr differed from the expected stderr.
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args lint/invalid_value-polymorphic.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/invalid_value-polymorphic.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/invalid_value-polymorphic" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--crate-type=lib" "-Zmir-enable-passes=+InstSimplify"
--- stderr -------------------------------
warning: MIR pass `InstSimplify` is unknown and will be ignored

warning: MIR pass `InstSimplify` is unknown and will be ignored
warning: MIR pass `InstSimplify` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `InstSimplify` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `InstSimplify` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `InstSimplify` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `InstSimplify` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: 6 warnings emitted
------------------------------------------



---- [ui] tests/ui/mir/alignment/misaligned-constant-gvn.rs stdout ----
Saved the actual stderr to "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/mir/alignment/misaligned-constant-gvn/misaligned-constant-gvn.stderr"
normalized stderr:
warning: MIR pass `GVN` is unknown and will be ignored

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: 12 warnings emitted





The actual stderr differed from the expected stderr.
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args mir/alignment/misaligned-constant-gvn.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/mir/alignment/misaligned-constant-gvn.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/mir/alignment/misaligned-constant-gvn" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Zmir-opt-level=0" "-Zmir-enable-passes=+GVN"
--- stderr -------------------------------
--- stderr -------------------------------
warning: MIR pass `GVN` is unknown and will be ignored

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: 12 warnings emitted
------------------------------------------



---- [ui] tests/ui/mir/debug-ref-undef.rs stdout ----
Saved the actual stderr to "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/mir/debug-ref-undef/debug-ref-undef.stderr"
normalized stderr:
warning: MIR pass `ReferencePropagation` is unknown and will be ignored

warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: 32 warnings emitted





The actual stderr differed from the expected stderr.
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args mir/debug-ref-undef.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/mir/debug-ref-undef.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/mir/debug-ref-undef/a" "-A" "internal_features" "-Crpath" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-g" "-O" "-Zmir-opt-level=0" "-Zinline-mir=y" "-Zmir-enable-passes=+ReferencePropagation"
--- stderr -------------------------------
warning: MIR pass `ReferencePropagation` is unknown and will be ignored

warning: MIR pass `ReferencePropagation` is unknown and will be ignored
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `ReferencePropagation` is unknown and will be ignored
   |
   |
---
To only update this specific test, also pass `--test-args mir/validate/validate-unsize-cast.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/mir/validate/validate-unsize-cast.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/mir/validate/validate-unsize-cast" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Zmir-opt-level=0" "-Zmir-enable-passes=+Inline,+GVN" "-Zvalidate-mir"
--- stderr -------------------------------
error[E0277]: the trait bound `T: Unsize<U>` is not satisfied
##[error]  --> /checkout/tests/ui/mir/validate/validate-unsize-cast.rs:10:42
   |
   |
LL | impl<T: ?Sized, U: ?Sized> CastTo<U> for T {}
   |                                          ^ the trait `Unsize<U>` is not implemented for `T`
   = note: all implementations of `Unsize` are provided automatically by the compiler, see <https://doc.rust-lang.org/stable/std/marker/trait.Unsize.html> for more information
note: required by a bound in `CastTo`
  --> /checkout/tests/ui/mir/validate/validate-unsize-cast.rs:7:30
   |
   |
LL | pub trait CastTo<U: ?Sized>: Unsize<U> {}
   |                              ^^^^^^^^^ required by this bound in `CastTo`
   |
   |
LL | impl<T: ?Sized + std::marker::Unsize<U>, U: ?Sized> CastTo<U> for T {}

warning: MIR pass `Inline` is unknown and will be ignored


warning: MIR pass `GVN` is unknown and will be ignored
warning: MIR pass `Inline` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `Inline` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `Inline` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `Inline` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `Inline` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `Inline` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `Inline` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `Inline` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `Inline` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `Inline` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `Inline` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
warning: MIR pass `Inline` is unknown and will be ignored
   |
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

warning: MIR pass `GVN` is unknown and will be ignored
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
error: aborting due to 1 previous error; 26 warnings emitted

For more information about this error, try `rustc --explain E0277`.
------------------------------------------
```

</details>

--- a/compiler/rustc_mir_transform/src/lib.rs
+++ b/compiler/rustc_mir_transform/src/lib.rs
@@ -40,77 +40,159 @@ use tracing::{debug, trace};
 #[macro_use]
 mod pass_manager;
 
+use std::sync::LazyLock;
+
 use pass_manager::{self as pm, Lint, MirLint, MirPass, WithMinOptLevel};
 
-mod abort_unwinding_calls;
-mod add_call_guards;
-mod add_moves_for_packed_drops;
-mod add_retag;
-mod add_subtyping_projections;
-mod check_alignment;
-mod check_const_item_mutation;
-mod check_packed_ref;
-mod check_undefined_transmutes;
-// This pass is public to allow external drivers to perform MIR cleanup
-pub mod cleanup_post_borrowck;
-mod copy_prop;
-mod coroutine;
 mod cost_checker;
-mod coverage;
 mod cross_crate_inline;
-mod ctfe_limit;
-mod dataflow_const_prop;
-mod dead_store_elimination;
 mod deduce_param_attrs;
-mod deduplicate_blocks;
-mod deref_separator;
-mod dest_prop;
-pub mod dump_mir;
-mod early_otherwise_branch;
-mod elaborate_box_derefs;
-mod elaborate_drops;
 mod errors;
 mod ffi_unwind_calls;
-mod function_item_references;
-mod gvn;
-// Made public so that `mir_drops_elaborated_and_const_checked` can be overridden
-// by custom rustc drivers, running all the steps by themselves. See #114628.
-pub mod inline;
-mod instsimplify;
-mod jump_threading;
-mod known_panics_lint;
-mod large_enums;
 mod lint;
-mod lower_intrinsics;
-mod lower_slice_len;
-mod match_branches;
-mod mentioned_items;
-mod multiple_return_terminators;
-mod nrvo;
-mod post_drop_elaboration;
-mod prettify;
-mod promote_consts;
-mod ref_prop;
-mod remove_noop_landing_pads;
-mod remove_place_mention;
-mod remove_storage_markers;
-mod remove_uninit_drops;
-mod remove_unneeded_drops;
-mod remove_zsts;
-mod required_consts;
-mod reveal_all;
-mod sanity_check;
 mod shim;
 mod ssa;
-// This pass is public to allow external drivers to perform MIR cleanup
-pub mod simplify;
-mod simplify_branches;
-mod simplify_comparison_integral;
-mod single_use_consts;
-mod sroa;
-mod unreachable_enum_branching;
-mod unreachable_prop;
-mod validate;
+
+/// We import passes via this macro so that we can have a static list of pass names
+/// (used to verify CLI arguments). It takes a list of modules, followed by the passes
+/// declared within them.
+/// ```ignore,macro-test
+/// declare_passes! {
+///     // Declare a single pass from the module `abort_unwinding_calls`
+///     mod abort_unwinding_calls : AbortUnwindingCalls;
+///     // When passes are grouped together as an enum, declare the two constituent passes
+///     mod add_call_guards : AddCallGuards {
+///         AllCallEdges,
+///         CriticalCallEdges
+///     };
+///     // Declares multiple pass groups, each containing their own constituent passes
+///     mod simplify : SimplifyCfg {
+///         Initial,
+///         /* omitted */
+///     }, SimplifyLocals {
+///         BeforeConstProp,
+///         /* omitted */
+///     };
+/// }
+/// ```
+macro_rules! declare_passes {
+    (
+        $(
+            $vis:vis mod $mod_name:ident : $($pass_name:ident $( { $($ident:ident),* } )?),+ $(,)?;
+        )*
+    ) => {
+        $(
+            $vis mod $mod_name;
+            $(
+                // Make sure the type name is correct
+                #[allow(unused_imports)]
+                use $mod_name::$pass_name as _;
+            )+
+        )*
+
+        #[cfg(debug_assertions)]
+        static PASS_NAMES: LazyLock<Vec<String>> = LazyLock::new(|| vec![
+            // Fake marker pass
+            "PreCodegen".to_string(),
+            $(
+                $(
+                    stringify!($pass_name).to_string(),
+                    $(
+                        $(
+                            $mod_name::$pass_name::$ident.name().to_string(),
+                        )*
+                    )?
+                )+
+            )*
+        ]);
+    };
+}
+
+declare_passes! {
+    mod abort_unwinding_calls : AbortUnwindingCalls;
+    mod add_call_guards : AddCallGuards { AllCallEdges, CriticalCallEdges };
+    mod add_moves_for_packed_drops : AddMovesForPackedDrops;
+    mod add_retag : AddRetag;
+    mod add_subtyping_projections : Subtyper;
+    mod check_alignment : CheckAlignment;
+    mod check_const_item_mutation : CheckConstItemMutation;
+    mod check_packed_ref : CheckPackedRef;
+    mod check_undefined_transmutes : CheckUndefinedTransmutes;
+    // This pass is public to allow external drivers to perform MIR cleanup
+    pub mod cleanup_post_borrowck : CleanupPostBorrowck;
+
+    mod copy_prop : CopyProp;
+    mod coroutine : StateTransform;
+    mod coverage : InstrumentCoverage;
+    mod ctfe_limit : CtfeLimit;
+    mod dataflow_const_prop : DataflowConstProp;
+    mod dead_store_elimination : DeadStoreElimination {
+        Initial,
+        Final
+    };
+    mod deduplicate_blocks : DeduplicateBlocks;
+    mod deref_separator : Derefer;
+    mod dest_prop : DestinationPropagation;
+    pub mod dump_mir : Marker;
+    mod early_otherwise_branch : EarlyOtherwiseBranch;
+    mod elaborate_box_derefs : ElaborateBoxDerefs;
+    mod elaborate_drops : ElaborateDrops;
+    mod function_item_references : FunctionItemReferences;
+    mod gvn : GVN;
+    // Made public so that `mir_drops_elaborated_and_const_checked` can be overridden
+    // by custom rustc drivers, running all the steps by themselves. See #114628.
+    pub mod inline : Inline;
+    mod instsimplify : InstSimplify { BeforeInline, AfterSimplifyCfg };
+    mod jump_threading : JumpThreading;
+    mod known_panics_lint : KnownPanicsLint;
+    mod large_enums : EnumSizeOpt;
+    mod lower_intrinsics : LowerIntrinsics;
+    mod lower_slice_len : LowerSliceLenCalls;
+    mod match_branches : MatchBranchSimplification;
+    mod mentioned_items : MentionedItems;
+    mod multiple_return_terminators : MultipleReturnTerminators;
+    mod nrvo : RenameReturnPlace;
+    mod post_drop_elaboration : CheckLiveDrops;
+    mod prettify : ReorderBasicBlocks, ReorderLocals;
+    mod promote_consts : PromoteTemps;
+    mod ref_prop : ReferencePropagation;
+    mod remove_noop_landing_pads : RemoveNoopLandingPads;
+    mod remove_place_mention : RemovePlaceMention;
+    mod remove_storage_markers : RemoveStorageMarkers;
+    mod remove_uninit_drops : RemoveUninitDrops;
+    mod remove_unneeded_drops : RemoveUnneededDrops;
+    mod remove_zsts : RemoveZsts;
+    mod required_consts : RequiredConstsVisitor;
+    mod reveal_all : RevealAll;
+    mod sanity_check : SanityCheck;
+    // This pass is public to allow external drivers to perform MIR cleanup
+    pub mod simplify :
+        SimplifyCfg {
+            Initial,
+            PromoteConsts,
+            RemoveFalseEdges,
+            PostAnalysis,
+            PreOptimizations,
+            Final,
+            MakeShim,
+            AfterUnreachableEnumBranching
+        },
+        SimplifyLocals {
+            BeforeConstProp,
+            AfterGVN,
+            Final
+        };
+    mod simplify_branches : SimplifyConstCondition {
+        AfterConstProp,
+        Final
+    };
+    mod simplify_comparison_integral : SimplifyComparisonIntegral;
+    mod single_use_consts : SingleUseConsts;
+    mod sroa : ScalarReplacementOfAggregates;
+    mod unreachable_enum_branching : UnreachableEnumBranching;
+    mod unreachable_prop : UnreachablePropagation;
+    mod validate : Validator;
+}
 
 rustc_fluent_macro::fluent_messages! { "../messages.ftl" }
 

--- a/compiler/rustc_mir_transform/src/lib.rs
+++ b/compiler/rustc_mir_transform/src/lib.rs
@@ -90,20 +90,20 @@ macro_rules! declare_passes {
             )+
         )*
 
-        static PASS_NAMES: LazyLock<Vec<String>> = LazyLock::new(|| vec![
+        static PASS_NAMES: LazyLock<FxIndexSet<&str>> = LazyLock::new(|| [
             // Fake marker pass
-            "PreCodegen".to_string(),
+            "PreCodegen",
             $(
                 $(
-                    stringify!($pass_name).to_string(),
+                    stringify!($pass_name),
                     $(
                         $(
-                            $mod_name::$pass_name::$ident.name().to_string(),
+                            $mod_name::$pass_name::$ident.name(),
                         )*
                     )?
                 )+
             )*
-        ]);
+        ].into_iter().collect());
     };
 }
 

--- a/compiler/rustc_mir_transform/src/pass_manager.rs
+++ b/compiler/rustc_mir_transform/src/pass_manager.rs
@@ -200,9 +200,8 @@ fn run_passes_inner<'tcx>(
 
     let named_passes: FxIndexSet<_> =
         overridden_passes.iter().map(|(name, _)| name.as_str()).collect();
-    let known_passes: FxIndexSet<_> = crate::PASS_NAMES.iter().map(|p| p.as_str()).collect();
 
-    for &name in named_passes.difference(&known_passes) {
+    for &name in named_passes.difference(&*crate::PASS_NAMES) {
         tcx.dcx().emit_warn(errors::UnknownPassName { name });
     }
 
@@ -213,7 +212,7 @@ fn run_passes_inner<'tcx>(
     {
         let used_passes: FxIndexSet<_> = passes.iter().map(|p| p.name()).collect();
 
-        let undeclared = used_passes.difference(&known_passes).collect::<Vec<_>>();
+        let undeclared = used_passes.difference(&*crate::PASS_NAMES).collect::<Vec<_>>();
         if let Some((name, rest)) = undeclared.split_first() {
             let mut err =
                 tcx.dcx().struct_bug(format!("pass `{name}` is not declared in `PASS_NAMES`"));

--- a/compiler/rustc_mir_transform/src/pass_manager.rs
+++ b/compiler/rustc_mir_transform/src/pass_manager.rs
@@ -1,7 +1,7 @@
 use std::cell::RefCell;
 use std::collections::hash_map::Entry;
 
-use rustc_data_structures::fx::FxHashMap;
+use rustc_data_structures::fx::{FxHashMap, FxIndexSet};
 use rustc_middle::mir::{self, Body, MirPhase, RuntimePhase};
 use rustc_middle::ty::TyCtxt;
 use rustc_session::Session;
@@ -197,6 +197,16 @@ fn run_passes_inner<'tcx>(
 ) {
     let overridden_passes = &tcx.sess.opts.unstable_opts.mir_enable_passes;
     trace!(?overridden_passes);
+
+    #[cfg(debug_assertions)]
+    {
+        let used_passes: FxIndexSet<_> = passes.iter().map(|p| p.name()).collect();
+        let known_passes: FxIndexSet<_> = crate::PASS_NAMES.iter().map(|p| p.as_str()).collect();
+
+        for &name in used_passes.difference(&known_passes) {
+            tcx.dcx().bug(format!("pass `{name}` is not declared in `PASS_NAMES`"));
+        }
+    }
 
     let prof_arg = tcx.sess.prof.enabled().then(|| format!("{:?}", body.source.def_id()));
 

--- a/compiler/rustc_mir_transform/src/pass_manager.rs
+++ b/compiler/rustc_mir_transform/src/pass_manager.rs
@@ -11,14 +11,15 @@ use crate::lint::lint_body;
 use crate::{errors, validate};
 
 thread_local! {
-    static PASS_NAMES: RefCell<FxHashMap<&'static str, &'static str>> = {
+    /// Maps MIR pass names to a snake case form to match profiling naming style
+    static PASS_TO_PROFILER_NAMES: RefCell<FxHashMap<&'static str, &'static str>> = {
         RefCell::new(FxHashMap::default())
     };
 }
 
 /// Converts a MIR pass name into a snake case form to match the profiling naming style.
 fn to_profiler_name(type_name: &'static str) -> &'static str {
-    PASS_NAMES.with(|names| match names.borrow_mut().entry(type_name) {
+    PASS_TO_PROFILER_NAMES.with(|names| match names.borrow_mut().entry(type_name) {
         Entry::Occupied(e) => *e.get(),
         Entry::Vacant(e) => {
             let snake_case: String = type_name

